### ISSUE TITLE
[spring-kafka] Connection to node -1 could not be established. Broker may not be available.

### ISSUE THREAD

**Issue by keltik85** [eugenp/tutorials#7197]
I changed the `kafka.bootstrapAddress` in the file `application.properties` to:
```
kafka.bootstrapAddress=ubuntu-kafka-vm:9092
```
On that remote address `ubuntu-kafka-vm:9092` I have a kafka installation running.

Unfortunately, when running the `KafkaApplication` in this tutorial I get the following warnings:

```

  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v2.1.6.RELEASE)

13:57:43.826 [main] INFO  c.b.spring.kafka.KafkaApplication - Starting KafkaApplication on mylaptop with PID 26948 (/home/myuser/eclipse-workspace/tutorials/spring-kafka/target/classes started by myuser in /home/myuser/eclipse-workspace/tutorials/spring-kafka)
13:57:43.827 [main] INFO  c.b.spring.kafka.KafkaApplication - No active profile set, falling back to default profiles: default
13:57:44.173 [main] INFO  o.s.c.s.PostProcessorRegistrationDelegate$BeanPostProcessorChecker - Bean 'org.springframework.kafka.annotation.KafkaBootstrapConfiguration' of type [org.springframework.kafka.annotation.KafkaBootstrapConfiguration$$EnhancerBySpringCGLIB$$18b91d20] is not eligible for getting processed by all BeanPostProcessors (for example: not eligible for auto-proxying)
13:57:44.538 [main] INFO  o.a.k.c.admin.AdminClientConfig - AdminClientConfig values: 
	bootstrap.servers = [192.168.56.102:9092]
	client.id = 
	connections.max.idle.ms = 300000
	metadata.max.age.ms = 300000
	metric.reporters = []
	metrics.num.samples = 2
	metrics.recording.level = INFO
	metrics.sample.window.ms = 30000
	receive.buffer.bytes = 65536
	reconnect.backoff.max.ms = 1000
	reconnect.backoff.ms = 50
	request.timeout.ms = 120000
	retries = 5
	retry.backoff.ms = 100
	sasl.client.callback.handler.class = null
	sasl.jaas.config = null
	sasl.kerberos.kinit.cmd = /usr/bin/kinit
	sasl.kerberos.min.time.before.relogin = 60000
	sasl.kerberos.service.name = null
	sasl.kerberos.ticket.renew.jitter = 0.05
	sasl.kerberos.ticket.renew.window.factor = 0.8
	sasl.login.callback.handler.class = null
	sasl.login.class = null
	sasl.login.refresh.buffer.seconds = 300
	sasl.login.refresh.min.period.seconds = 60
	sasl.login.refresh.window.factor = 0.8
	sasl.login.refresh.window.jitter = 0.05
	sasl.mechanism = GSSAPI
	security.protocol = PLAINTEXT
	send.buffer.bytes = 131072
	ssl.cipher.suites = null
	ssl.enabled.protocols = [TLSv1.2, TLSv1.1, TLSv1]
	ssl.endpoint.identification.algorithm = https
	ssl.key.password = null
	ssl.keymanager.algorithm = SunX509
	ssl.keystore.location = null
	ssl.keystore.password = null
	ssl.keystore.type = JKS
	ssl.protocol = TLS
	ssl.provider = null
	ssl.secure.random.implementation = null
	ssl.trustmanager.algorithm = PKIX
	ssl.truststore.location = null
	ssl.truststore.password = null
	ssl.truststore.type = JKS

13:57:44.563 [main] INFO  o.a.kafka.common.utils.AppInfoParser - Kafka version : 2.0.1
13:57:44.563 [main] INFO  o.a.kafka.common.utils.AppInfoParser - Kafka commitId : fa14705e51bd2ce5
13:57:44.652 [main] INFO  o.a.k.c.consumer.ConsumerConfig - ConsumerConfig values: 
	auto.commit.interval.ms = 5000
	auto.offset.reset = latest
	bootstrap.servers = [192.168.56.102:9092]
	check.crcs = true
	client.id = 
	connections.max.idle.ms = 540000
	default.api.timeout.ms = 60000
	enable.auto.commit = true
	exclude.internal.topics = true
	fetch.max.bytes = 52428800
	fetch.max.wait.ms = 500
	fetch.min.bytes = 1
	group.id = headers
	heartbeat.interval.ms = 3000
	interceptor.classes = []
	internal.leave.group.on.close = true
	isolation.level = read_uncommitted
	key.deserializer = class org.apache.kafka.common.serialization.StringDeserializer
	max.partition.fetch.bytes = 1048576
	max.poll.interval.ms = 300000
	max.poll.records = 500
	metadata.max.age.ms = 300000
	metric.reporters = []
	metrics.num.samples = 2
	metrics.recording.level = INFO
	metrics.sample.window.ms = 30000
	partition.assignment.strategy = [class org.apache.kafka.clients.consumer.RangeAssignor]
	receive.buffer.bytes = 65536
	reconnect.backoff.max.ms = 1000
	reconnect.backoff.ms = 50
	request.timeout.ms = 30000
	retry.backoff.ms = 100
	sasl.client.callback.handler.class = null
	sasl.jaas.config = null
	sasl.kerberos.kinit.cmd = /usr/bin/kinit
	sasl.kerberos.min.time.before.relogin = 60000
	sasl.kerberos.service.name = null
	sasl.kerberos.ticket.renew.jitter = 0.05
	sasl.kerberos.ticket.renew.window.factor = 0.8
	sasl.login.callback.handler.class = null
	sasl.login.class = null
	sasl.login.refresh.buffer.seconds = 300
	sasl.login.refresh.min.period.seconds = 60
	sasl.login.refresh.window.factor = 0.8
	sasl.login.refresh.window.jitter = 0.05
	sasl.mechanism = GSSAPI
	security.protocol = PLAINTEXT
	send.buffer.bytes = 131072
	session.timeout.ms = 10000
	ssl.cipher.suites = null
	ssl.enabled.protocols = [TLSv1.2, TLSv1.1, TLSv1]
	ssl.endpoint.identification.algorithm = https
	ssl.key.password = null
	ssl.keymanager.algorithm = SunX509
	ssl.keystore.location = null
	ssl.keystore.password = null
	ssl.keystore.type = JKS
	ssl.protocol = TLS
	ssl.provider = null
	ssl.secure.random.implementation = null
	ssl.trustmanager.algorithm = PKIX
	ssl.truststore.location = null
	ssl.truststore.password = null
	ssl.truststore.type = JKS
	value.deserializer = class org.apache.kafka.common.serialization.StringDeserializer

13:57:44.677 [main] INFO  o.a.kafka.common.utils.AppInfoParser - Kafka version : 2.0.1
13:57:44.677 [main] INFO  o.a.kafka.common.utils.AppInfoParser - Kafka commitId : fa14705e51bd2ce5
13:57:44.683 [main] INFO  org.apache.kafka.clients.Metadata - Cluster ID: F9VdK8e1Txit9tEahocvzw
13:57:44.697 [main] INFO  o.a.k.c.consumer.ConsumerConfig - ConsumerConfig values: 
	auto.commit.interval.ms = 5000
	auto.offset.reset = latest
	bootstrap.servers = [192.168.56.102:9092]
	check.crcs = true
	client.id = 
	connections.max.idle.ms = 540000
	default.api.timeout.ms = 60000
	enable.auto.commit = true
	exclude.internal.topics = true
	fetch.max.bytes = 52428800
	fetch.max.wait.ms = 500
	fetch.min.bytes = 1
	group.id = headers
	heartbeat.interval.ms = 3000
	interceptor.classes = []
	internal.leave.group.on.close = true
	isolation.level = read_uncommitted
	key.deserializer = class org.apache.kafka.common.serialization.StringDeserializer
	max.partition.fetch.bytes = 1048576
	max.poll.interval.ms = 300000
	max.poll.records = 500
	metadata.max.age.ms = 300000
	metric.reporters = []
	metrics.num.samples = 2
	metrics.recording.level = INFO
	metrics.sample.window.ms = 30000
	partition.assignment.strategy = [class org.apache.kafka.clients.consumer.RangeAssignor]
	receive.buffer.bytes = 65536
	reconnect.backoff.max.ms = 1000
	reconnect.backoff.ms = 50
	request.timeout.ms = 30000
	retry.backoff.ms = 100
	sasl.client.callback.handler.class = null
	sasl.jaas.config = null
	sasl.kerberos.kinit.cmd = /usr/bin/kinit
	sasl.kerberos.min.time.before.relogin = 60000
	sasl.kerberos.service.name = null
	sasl.kerberos.ticket.renew.jitter = 0.05
	sasl.kerberos.ticket.renew.window.factor = 0.8
	sasl.login.callback.handler.class = null
	sasl.login.class = null
	sasl.login.refresh.buffer.seconds = 300
	sasl.login.refresh.min.period.seconds = 60
	sasl.login.refresh.window.factor = 0.8
	sasl.login.refresh.window.jitter = 0.05
	sasl.mechanism = GSSAPI
	security.protocol = PLAINTEXT
	send.buffer.bytes = 131072
	session.timeout.ms = 10000
	ssl.cipher.suites = null
	ssl.enabled.protocols = [TLSv1.2, TLSv1.1, TLSv1]
	ssl.endpoint.identification.algorithm = https
	ssl.key.password = null
	ssl.keymanager.algorithm = SunX509
	ssl.keystore.location = null
	ssl.keystore.password = null
	ssl.keystore.type = JKS
	ssl.protocol = TLS
	ssl.provider = null
	ssl.secure.random.implementation = null
	ssl.trustmanager.algorithm = PKIX
	ssl.truststore.location = null
	ssl.truststore.password = null
	ssl.truststore.type = JKS
	value.deserializer = class org.apache.kafka.common.serialization.StringDeserializer

13:57:44.703 [main] INFO  o.a.kafka.common.utils.AppInfoParser - Kafka version : 2.0.1
13:57:44.703 [main] INFO  o.a.kafka.common.utils.AppInfoParser - Kafka commitId : fa14705e51bd2ce5
13:57:44.705 [main] INFO  o.s.s.c.ThreadPoolTaskScheduler - Initializing ExecutorService
13:57:44.709 [main] INFO  o.a.k.c.consumer.ConsumerConfig - ConsumerConfig values: 
	auto.commit.interval.ms = 5000
	auto.offset.reset = latest
	bootstrap.servers = [localhost:9092]
	check.crcs = true
	client.id = 
	connections.max.idle.ms = 540000
	default.api.timeout.ms = 60000
	enable.auto.commit = true
	exclude.internal.topics = true
	fetch.max.bytes = 52428800
	fetch.max.wait.ms = 500
	fetch.min.bytes = 1
	group.id = 
	heartbeat.interval.ms = 3000
	interceptor.classes = []
	internal.leave.group.on.close = true
	isolation.level = read_uncommitted
	key.deserializer = class org.apache.kafka.common.serialization.StringDeserializer
	max.partition.fetch.bytes = 1048576
	max.poll.interval.ms = 300000
	max.poll.records = 500
	metadata.max.age.ms = 300000
	metric.reporters = []
	metrics.num.samples = 2
	metrics.recording.level = INFO
	metrics.sample.window.ms = 30000
	partition.assignment.strategy = [class org.apache.kafka.clients.consumer.RangeAssignor]
	receive.buffer.bytes = 65536
	reconnect.backoff.max.ms = 1000
	reconnect.backoff.ms = 50
	request.timeout.ms = 30000
	retry.backoff.ms = 100
	sasl.client.callback.handler.class = null
	sasl.jaas.config = null
	sasl.kerberos.kinit.cmd = /usr/bin/kinit
	sasl.kerberos.min.time.before.relogin = 60000
	sasl.kerberos.service.name = null
	sasl.kerberos.ticket.renew.jitter = 0.05
	sasl.kerberos.ticket.renew.window.factor = 0.8
	sasl.login.callback.handler.class = null
	sasl.login.class = null
	sasl.login.refresh.buffer.seconds = 300
	sasl.login.refresh.min.period.seconds = 60
	sasl.login.refresh.window.factor = 0.8
	sasl.login.refresh.window.jitter = 0.05
	sasl.mechanism = GSSAPI
	security.protocol = PLAINTEXT
	send.buffer.bytes = 131072
	session.timeout.ms = 10000
	ssl.cipher.suites = null
	ssl.enabled.protocols = [TLSv1.2, TLSv1.1, TLSv1]
	ssl.endpoint.identification.algorithm = https
	ssl.key.password = null
	ssl.keymanager.algorithm = SunX509
	ssl.keystore.location = null
	ssl.keystore.password = null
	ssl.keystore.type = JKS
	ssl.protocol = TLS
	ssl.provider = null
	ssl.secure.random.implementation = null
	ssl.trustmanager.algorithm = PKIX
	ssl.truststore.location = null
	ssl.truststore.password = null
	ssl.truststore.type = JKS
	value.deserializer = class org.apache.kafka.common.serialization.StringDeserializer

13:57:44.714 [main] INFO  o.a.kafka.common.utils.AppInfoParser - Kafka version : 2.0.1
13:57:44.715 [main] INFO  o.a.kafka.common.utils.AppInfoParser - Kafka commitId : fa14705e51bd2ce5
13:57:44.718 [main] WARN  o.apache.kafka.clients.NetworkClient - [Consumer clientId=consumer-3, groupId=] Connection to node -1 could not be established. Broker may not be available.
13:57:44.724 [org.springframework.kafka.KafkaListenerEndpointContainer#2-0-C-1] INFO  org.apache.kafka.clients.Metadata - Cluster ID: F9VdK8e1Txit9tEahocvzw
13:57:44.724 [org.springframework.kafka.KafkaListenerEndpointContainer#2-0-C-1] INFO  o.a.k.c.c.i.AbstractCoordinator - [Consumer clientId=consumer-2, groupId=headers] Discovered group coordinator ubuntu-kafka-vm:9092 (id: 2147483647 rack: null)
13:57:44.726 [org.springframework.kafka.KafkaListenerEndpointContainer#2-0-C-1] INFO  o.a.k.c.c.i.ConsumerCoordinator - [Consumer clientId=consumer-2, groupId=headers] Revoking previously assigned partitions []
13:57:44.726 [org.springframework.kafka.KafkaListenerEndpointContainer#2-0-C-1] INFO  o.s.k.l.KafkaMessageListenerContainer - partitions revoked: []
13:57:44.726 [org.springframework.kafka.KafkaListenerEndpointContainer#2-0-C-1] INFO  o.a.k.c.c.i.AbstractCoordinator - [Consumer clientId=consumer-2, groupId=headers] (Re-)joining group
13:57:44.819 [main] WARN  o.apache.kafka.clients.NetworkClient - [Consumer clientId=consumer-3, groupId=] Connection to node -1 could not be established. Broker may not be available.
13:57:45.021 [main] WARN  o.apache.kafka.clients.NetworkClient - [Consumer clientId=consumer-3, groupId=] Connection to node -1 could not be established. Broker may not be available.
13:57:45.223 [main] WARN  o.apache.kafka.clients.NetworkClient - [Consumer clientId=consumer-3, groupId=] Connection to node -1 could not be established. Broker may not be available.
13:57:45.627 [main] WARN  o.apache.kafka.clients.NetworkClient - [Consumer clientId=consumer-3, groupId=] Connection to node -1 could not be established. Broker may not be available.
13:57:46.434 [main] WARN  o.apache.kafka.clients.NetworkClient - [Consumer clientId=consumer-3, groupId=] Connection to node -1 could not be established. Broker may not be available.

```
I do not get the messages that are described on the [README.md](https://github.com/eugenp/tutorials/tree/master/spring-kafka#message-received-from-the-baeldung-topic-by-the-basic-listeners-with-groups-foo-and-bar). What is wrong with the configuration?

I can check that the kafka installation on that server is working correctly. From my laptops bash I can interact with the kafka installation in the following fashion without running into any kind of errors:

```
myuser@mylaptop:/tmp$ bin/kafka-topics.sh --create --zookeeper 192.168.56.102:2181 --replication-factor 1 --partitions 1 --topic HundrethTopic
Created topic HundrethTopic.

myuser@mylaptop:/tmp$ bin/kafka-console-consumer.sh --bootstrap-server 192.168.56.102:9092 --topic HundrethTopic --from-beginning
[comment: recieves everything send by the next bash command]

myuser@mylaptop:/tmp$ bin/kafka-console-producer.sh --broker-list 192.168.56.102:9092 --topic HundrethTopic
[comment: sends everything to any subscribed consumer, only one in this example, see above bash command]
```

So in other words "Creating a topic", "Producing" and "Consuming" work as advertised according to the Getting started part in the [Apache Kafka Documentation](https://kafka.apache.org/quickstart) using my bash.

**Comment by maibin** [comment]
@keltik85 could you please let us know to which article you are referring to?

**Comment by keltik85** [comment]
> @keltik85 could you please let us know to which article you are referring to?

It is the article you can read in the heading portion of this issue [spring-kafka] and the webpage is https://www.baeldung.com/spring-kafka .

**Comment by vikasramireddy** [comment]
Kafka is not started in your machine. Please go through https://kafka.apache.org/quickstart

**Comment by maibin** [comment]
@keltik85 @vikasramireddy we logged an internal ticket to fix this issue. I'll follow up once it is resolved. Thanks!

**Comment by lor6** [comment]
Quick recap:

- all the custom ContainerFactories in KafkaConsumerConfig.java set the broker connection using the kafka.bootstrapAddress property
- since this is a spring-boot project it also provides an auto-configuration that sets up a default ContainerFactory
- the defaults use the spring.kafka.* properties to customize how autoconfiguration works
- the listenToParition KafkaListener is not using the defined partitionsKafkaListenerContainerFactory and it falls back to the default one, provided by auto-configuration (this one always tries to connect by default to localhost:9092 unless you actually override its properties)
- and because of this the clients spawned by this listener fail to connect to the broker

Fixed in https://github.com/eugenp/tutorials/pull/8782
